### PR TITLE
MC-11 Settings presentation

### PR DESCRIPTION
**Pull request by bugbenderx**

### CODE REVIEW

File: presentation/src/main/java/com/example/presentation/settings/CurrencyUi.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import com.example.presentation.core.views.ChangeText
+import com.example.presentation.core.views.ChangeVisibility
+
+data class CurrencyUi(private val value: String, private val chosen: Boolean = false) {
+
+    fun areItemsTheSame(currencyUi: CurrencyUi) = currencyUi.value == this.value
+
+    fun show(currencyTextView: ChangeText, chosenImageView: ChangeVisibility) {
+        currencyTextView.changeText(value)
+        chosenImageView.also { if (chosen) it.show() else it.hide() }
```
Comment (JohnnySC) [inline presentation/src/main/java/com/example/presentation/settings/CurrencyUi.kt]: run

File: presentation/src/main/java/com/example/presentation/settings/adapter/CurrencyAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import androidx.recyclerview.widget.RecyclerView
+import com.example.presentation.core.UpdateAdapter
+import com.example.presentation.databinding.ViewholderCurrencyBinding
+import com.example.presentation.settings.CurrencyUi
+
+class CurrencyAdapter(
+    private val chooseCurrencyBlock: (String) -> Unit
+) : RecyclerView.Adapter<CurrencyAdapter.CurrencyViewHolder>(), UpdateAdapter<CurrencyUi> {
+
+    private val currencies = mutableListOf<CurrencyUi>()
```
Comment (JohnnySC) [inline presentation/src/main/java/com/example/presentation/settings/adapter/CurrencyAdapter.kt:18]: fun selected() : String = currencies.find {it.isselected}?:""

File: presentation/src/main/java/com/example/presentation/settings/SettingsUiState.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package com.example.presentation.settings
+
+import com.example.presentation.core.UpdateAdapter
+
+interface SettingsUiState {
+
+    fun show(
```
Comment (JohnnySC) [inline presentation/src/main/java/com/example/presentation/settings/SettingsUiState.kt:8]: fun show(saveButton)

File: presentation/src/main/java/com/example/presentation/settings/adapter/CurrencyAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.RecyclerView
+import com.example.presentation.core.UpdateAdapter
+import com.example.presentation.databinding.ViewholderCurrencyBinding
+import com.example.presentation.settings.CurrencyUi
+
+class CurrencyAdapter(
+    private val chooseCurrencyBlock: (String) -> Unit
```
Comment (JohnnySC) [inline presentation/src/main/java/com/example/presentation/settings/adapter/CurrencyAdapter.kt:15]: 2 types

File: presentation/src/main/res/layout/fragment_settings.xml [hunk 5]
```diff
@@ -1,13 +1,51 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
     android:layout_width="match_parent"
     android:layout_height="match_parent"
     android:gravity="center"
     android:orientation="vertical">
 
-    <TextView
+    <ImageButton
         android:layout_width="wrap_content"
```
Comment (JohnnySC) [inline presentation/src/main/res/layout/fragment_settings.xml:12]: id

File: presentation/src/main/res/layout/fragment_settings.xml [hunk 6]
```diff
@@ -1,13 +1,51 @@
...
+        android:weightSum="2">
+
+        <androidx.recyclerview.widget.RecyclerView
+            android:id="@+id/fromCurrencyRecyclerView"
+            android:layout_width="0dp"
+            android:layout_height="match_parent"
+            android:layout_weight="1"
+            app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
+            tools:listitem="@layout/viewholder_currency" />
+
```
Comment (JohnnySC) [inline presentation/src/main/res/layout/fragment_settings.xml:33]: <View
background = color/black
width = 1dp
height = match_parent

**Review by JohnnySC — CHANGES_REQUESTED** [review]

File: presentation/src/main/java/com/example/presentation/settings/SettingsViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        communication.updateUi(
+            SettingsUiState.ReadyToSave(
+                toCurrencies = availableDestinations.map {
+                    CurrencyUi(value = it, chosen = it == currency)
+                }
+            )
+        )
+    }
+
+    fun save() = runAsync({
```
Comment (JohnnySC) [inline presentation/src/main/java/com/example/presentation/settings/SettingsViewModel.kt]: fun save(from:String, to: String)

**Review by JohnnySC — CHANGES_REQUESTED** [review]

File: presentation/src/main/java/com/example/presentation/settings/CurrencyUi.kt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            currencyTextView.changeText(value)
+            chosenImageView.run { if (chosen) show() else hide() }
+        }
+
+        override fun isSelected() = chosen
+
+        override fun value() = value
+
+        override fun areItemsTheSame(currencyUi: CurrencyUi) =
+            (currencyUi as? Base)?.value == this.value
```
Comment (JohnnySC) [inline presentation/src/main/java/com/example/presentation/settings/CurrencyUi.kt]: no need to cast: use value()

File: presentation/src/main/java/com/example/presentation/settings/SettingsViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    fun chooseFrom(currency: String) = runAsync({
+        val currencies = repository.availableDestinations(currency)
+        SettingsUiState.AvailableDestinations(
+            fromCurrencies = repository.allCurrencies().map {
+                CurrencyUi.Base(value = it, chosen = it == currency)
+            },
+            toCurrencies = currencies.run {
+                if (isEmpty()) listOf(CurrencyUi.Empty)
```
Comment (JohnnySC) [inline presentation/src/main/java/com/example/presentation/settings/SettingsViewModel.kt]: maybe mapper?

**Review by JohnnySC — CHANGES_REQUESTED** [review]

File: presentation/src/main/java/com/example/presentation/settings/CurrencyUi.kt (hunk 8 above)
Comment (bugbenderx) [reply presentation/src/main/java/com/example/presentation/settings/CurrencyUi.kt]: fixed

File: presentation/src/main/java/com/example/presentation/settings/SettingsViewModel.kt (hunk 9 above)
Comment (bugbenderx) [reply presentation/src/main/java/com/example/presentation/settings/SettingsViewModel.kt]: fixed

File: presentation/src/main/java/com/example/presentation/core/UpdateAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package com.example.presentation.core
+
+interface UpdateAdapter<T> {
```
Comment (JohnnySC) [inline presentation/src/main/java/com/example/presentation/core/UpdateAdapter.kt]: T: Any

**Review by JohnnySC — CHANGES_REQUESTED** [review]

**Review by JohnnySC — APPROVED** [review]

**Pull request merged**